Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp

```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
 
-void WebDriverBidiProcessor::navigate(const BrowsingContext& browsingContext, const String& url, std::optional<Inspector::Protocol::BidiBrowsingContext::ReadinessState>&&, CommandCallbackOf<String, String>&& callback)
+void WebDriverBidiProcessor::activate(const BrowsingContext& browsingContext, CommandCallback<void>&& callback)
 {
     RefPtr session = m_session.get();
     if (!session)
         ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
 
-    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
+    RefPtr webPageProxy = session->webPageProxyForHandle(browsingContext);
+    if (!webPageProxy)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(WindowNotFound);
+
+    // FIXME: detect non-top level browsing contexts, returning `invalid argument`.
+
+    session->switchToBrowsingContext(browsingContext, emptyString(), [callback = WTFMove(callback)](CommandResult<void>&& result) {
+        if (!result) {
+            callback(makeUnexpected(result.error()));
+            return;
+        }
+
+        callback({ });
+    });
+}
+
+void WebDriverBidiProcessor::close(const BrowsingContext& browsingContext, std::optional<bool>&& optionalPromptUnload, CommandCallback<void>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `promptUnload` option.
+    // FIXME: raise `invalid argument` if `browsingContext` is not a top-level traversable.
+
+    session->closeBrowsingContext(browsingContext);
+
+    callback({ });
+}
+
+void WebDriverBidiProcessor::create(Inspector::Protocol::BidiBrowsingContext::CreateType requestedCreateType, const BrowsingContext& optionalReferenceContext, std::optional<bool>&& optionalBackground, const String& optionalUserContext, CommandCallback<BrowsingContext>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `referenceContext` option.
+    // FIXME: implement `background` option.
+    // FIXME: implement `userContext` option.
+
+    auto createType = Inspector::Protocol::BidiBrowsingContext::CreateType::Tab;
+    if (requestedCreateType != createType)
+        createType = Inspector::Protocol::BidiBrowsingContext::CreateType::Window;
+
+    auto presentationHint = createType == Inspector::Protocol::BidiBrowsingContext::CreateType::Tab ? Inspector::Protocol::Automation::BrowsingContextPresentation::Tab : Inspector::Protocol::Automation::BrowsingContextPresentation::Window;
+    session->createBrowsingContext(presentationHint, [callback = WTFMove(callback)](CommandResultOf<BrowsingContext, Inspector::Protocol::Automation::BrowsingContextPresentation>&& result) {
+        if (!result) {
+            callback(makeUnexpected(result.error()));
+            return;
+        }
+
+        auto [resultContext, resultPresentation] = WTFMove(result.value());
+        callback(resultContext);
```

Comment:
should this be a `WTFMove(resultContext)` since it's moved out of `result.value()`?